File: docs/site/versioned_examples/version-21.09.0/python/acknowledgements.py
import asapo_consumer
import asapo_producer

def callback(payload,err):
    if err is not None and not isinstance(err, asapo_producer.AsapoServerWarning):
        print("could not send: ",payload,err)
    elif err is not None:
        print("sent with warning: ",payload,err)
    else:
        print("successfuly sent: ",payload)

endpoint = "localhost:8400"
beamtime = "asapo_test"

token = str("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e"
"yJleHAiOjk1NzE3MTAyMTYsImp0aSI6ImMzaXFhbGpmNDNhbGZ"
"wOHJua20wIiwic3ViIjoiYnRfYXNhcG9fdGVzdCIsIkV4dHJhQ"
"2xhaW1zIjp7IkFjY2Vzc1R5cGVzIjpbIndyaXRlIiwicmVhZCJ"
"dfX0.dkWupPO-ysI4t-jtWiaElAzDyJF6T7hu_Wz_Au54mYU")

path_to_files = "/var/tmp/asapo/global_shared/data/test_facility/gpfs/test/2019/data/asapo_test"

producer = asapo_producer.create_producer(endpoint, 'processed', beamtime, 'auto', 'test_source', '', 1, 60000)
producer.set_log_level('error')

# let's start with producing a sample of 10 simple messages
for i in range(1, 11):
    producer.send(i, "processed/test_file_ack_" + str(i), ('message #' + str(i)).encode(), stream = "default", callback = callback)

producer.wait_requests_finished(2000)

consumer = asapo_consumer.create_consumer(endpoint, path_to_files, True, beamtime, "test_source", token, 5000)
group_id = consumer.generate_group_id()

# the flag to separate the first attempt for message #3
firstTryNegative = True

try:
    while True:
        data, meta = consumer.get_next(group_id, meta_only = False)
        text_data = data.tobytes().decode("utf-8")
        message_id = meta['_id']

        # acknowledge all the messages except these
        if message_id not in [3,5,7]:
            print('Acknowledge the message #', message_id)
            consumer.acknowledge(group_id, message_id)

        # for message #3 we issue a negative acknowledgement, which will put it at the next place in the stream
        # in this case, it will be put in the end of a stream
        if message_id == 3:
            if firstTryNegative:
                print('Negative acknowledgement of the message #', message_id)
                # make the acknowledgement with a delay of 1 second
                consumer.neg_acknowledge(group_id, message_id, delay_ms=2000)
                firstTryNegative = False
            else:
                # on our second attempt we acknowledge the message
                print('Second try of the message #', message_id)
                consumer.acknowledge(group_id, message_id)

except asapo_consumer.AsapoStreamFinishedError:
    print('stream finished')

except asapo_consumer.AsapoEndOfStreamError:
    print('stream ended')

for message_id in consumer.get_unacknowledged_messages(group_id):
    data, meta = consumer.get_by_id(message_id, meta_only = False)
    print('Unacknowledged message:', data.tobytes().decode("utf-8"), meta)

File: docs/site/examples/cpp/acknowledgements.cpp
#include "asapo/asapo_producer.h"
#include "asapo/asapo_consumer.h"
#include <iostream>
#include <set>

void ProcessAfterSend(asapo::RequestCallbackPayload payload, asapo::Error err) {
    if (err && err != asapo::ProducerErrorTemplates::kServerWarning) {
        std::cerr << "error during send: " << err << std::endl;
        return;
    } else if (err) {
        std::cout << "warning during send: " << err << std::endl;
    } else {
        std::cout << "successfuly send " << payload.original_header.Json() << std::endl;
        return;
    }
}

void exit_if_error(std::string error_string, const asapo::Error& err) {
    if (err) {
        std::cerr << error_string << err << std::endl;
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char* argv[]) {
    asapo::Error err;

    auto endpoint = "localhost:8400";
    auto beamtime = "asapo_test";

    auto token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJl"
                 "eHAiOjk1NzE3MTAyMTYsImp0aSI6ImMzaXFhbGpmN"
                 "DNhbGZwOHJua20wIiwic3ViIjoiYnRfYXNhcG9fdG"
                 "VzdCIsIkV4dHJhQ2xhaW1zIjp7IkFjY2Vzc1R5cGV"
                 "zIjpbIndyaXRlIiwicmVhZCJdfX0.dkWupPO-ysI4"
                 "t-jtWiaElAzDyJF6T7hu_Wz_Au54mYU";

    auto path_to_files = "/var/tmp/asapo/global_shared/data/test_facility/gpfs/test/2019/data/asapo_test";

    auto credentials = asapo::SourceCredentials{asapo::SourceType::kProcessed, beamtime, "", "test_source", token};

    auto producer = asapo::Producer::Create(endpoint, 1, asapo::RequestHandlerType::kTcp, credentials, 60000, &err);
    exit_if_error("Cannot start producer", err);

    producer->SetLogLevel(asapo::LogLevel::Error);

    err = producer->DeleteStream("default", 1000, asapo::DeleteStreamOptions{true, true});
    exit_if_error("Cannot delete stream", err);

    // let's start with producing a sample of 10 simple messages
    for (uint64_t i = 1; i <= 10; i++) {
        std::string to_send = "message#" + std::to_string(i);
        auto send_size = to_send.size() + 1;
        auto buffer =  asapo::MessageData(new uint8_t[send_size]);
        memcpy(buffer.get(), to_send.c_str(), send_size);

        asapo::MessageHeader message_header{i, send_size, "processed/test_file_" + std::to_string(i)};
        err = producer->Send(message_header, std::move(buffer), asapo::kDefaultIngestMode, "default", &ProcessAfterSend);
        exit_if_error("Cannot send message", err);
    }

    err = producer->WaitRequestsFinished(2000);
    exit_if_error("Producer exit on timeout", err);

    auto consumer = asapo::ConsumerFactory::CreateConsumer(endpoint, path_to_files, true, credentials, &err);
    exit_if_error("Cannot start consumer", err);
    consumer->SetTimeout(5000);
    auto group_id = consumer->GenerateNewGroupId(&err);
    exit_if_error("Cannot create group id", err);

    // consume snippet_start
    asapo::MessageMeta mm;
    asapo::MessageData data;

    const std::set<int> ids {3, 5, 7};

    // the flag to separate the first attempt for message #3
    bool firstTryNegative = true;

    do {
        err = consumer->GetNext(group_id, &mm, &data, "default");

        if (err && err == asapo::ConsumerErrorTemplates::kStreamFinished) {
            std::cout << "stream finished" << std::endl;
            break;
        }

        if (err && err == asapo::ConsumerErrorTemplates::kEndOfStream) {
            std::cout << "stream ended" << std::endl;
            break;
        }
        exit_if_error("Cannot get next record", err); // snippet_end_remove

        // acknowledge all the messages except the ones in the set
        if (ids.find(mm.id) == ids.end()) {
            std::cout << "Acknowledge the message #" << mm.id << std::endl;
            consumer->Acknowledge(group_id, mm.id, "default");
        }

        // for message #3 we issue a negative acknowledgement, which will put it at the next place in the stream
        // in this case, it will be put in the end of a stream
        if (mm.id == 3) {
            if (firstTryNegative) {
                std::cout << "Negative acknowledgement of the message #" << mm.id << std::endl;
                // make the acknowledgement with a delay of 1 second
                consumer->NegativeAcknowledge(group_id, mm.id, 2000, "default");
                firstTryNegative = false;
            } else {
                // on our second attempt we acknowledge the message
                std::cout << "Second try of the message #" << mm.id << std::endl;
                consumer->Acknowledge(group_id, mm.id, "default");
            }
        }
    } while (1);
    // consume snippet_end

    // print snippet_start
    auto unacknowledgedMessages = consumer->GetUnacknowledgedMessages(group_id, 0, 0, "default", &err);
    exit_if_error("Could not get list of messages", err); // snippet_end_remove

    for (int i = 0; i < unacknowledgedMessages.size(); i++) {
        err = consumer->GetById(unacknowledgedMessages[i], &mm, &data, "default");
        exit_if_error("Cannot get message", err); // snippet_end_remove

        std::cout << "Unacknowledged message: " << reinterpret_cast<char const*>(data.get()) << std::endl;
        std::cout << "id: " << mm.id << std::endl;
        std::cout << "file name: " << mm.name << std::endl;
    }
    // print snippet_end

   return EXIT_SUCCESS;
}
